### PR TITLE
[12.0][MIG] fieldservice_recurring

### PR DESCRIPTION
**Pull request by kittiu**
Done migration and test script
#15

### CODE REVIEW

**Comment by brian10048** [comment]
Thank you!

It is missing the ability to install the module from base fieldservice settings

See example from v11:

https://github.com/OCA/field-service/blob/60c30bdd84f61222d7b0a7905c346220ea7fee23/fieldservice/models/res_config_settings.py#L43-L44

https://github.com/OCA/field-service/blob/60c30bdd84f61222d7b0a7905c346220ea7fee23/fieldservice/views/res_config_settings.xml#L106-L116

**Review by max3903 — APPROVED** [review]
LGTM. No test.

**Comment by brian10048** [comment]
I tested on the OCA runbot build and had some issues.

After installing fieldservice_recurring

### Steps to Reproduce:
Go to Field Service app.
Navigate to menu Configuration > Orders > select Frequencies

### Receive Error:
```
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 904, in search_read
    return self.do_search_read(model, fields, offset, limit, domain, sort)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 926, in do_search_read
    offset=offset or 0, limit=limit or False, order=sort or False)
  File "/.repo_requirements/odoo/odoo/models.py", line 4572, in search_read
    records = self.search(domain or [], offset=offset, limit=limit, order=order)
  File "/.repo_requirements/odoo/odoo/models.py", line 1561, in search
    res = self._search(args, offset=offset, limit=limit, order=order, count=count)
  File "/.repo_requirements/odoo/odoo/models.py", line 4105, in _search
    self._apply_ir_rules(query, 'read')
  File "/.repo_requirements/odoo/odoo/models.py", line 3957, in _apply_ir_rules
    where_clause, where_params, tables = Rule.domain_get(self._name, mode)
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_rule.py", line 126, in domain_get
    query = self.env[model_name].sudo()._where_calc(dom, active_test=False)
  File "/.repo_requirements/odoo/odoo/models.py", line 3896, in _where_calc
    e = expression.expression(domain, self)
  File "/.repo_requirements/odoo/odoo/osv/expression.py", line 672, in __init__
    self.parse()
  File "/.repo_requirements/odoo/odoo/osv/expression.py", line 853, in parse
    raise ValueError("Invalid field %r in leaf %r" % (left, str(leaf)))
ValueError: Invalid field 'company_id' in leaf "<osv.ExtendedLeaf: ('company_id', '=', False) on fsm_frequency (ctx: )>"
```

### Expected Result:
The Frequency view is loaded

**Comment by brian10048** [comment]
Additionally, from the OCA runbot build

The Configuration > Orders menu does not show the Frequency Set menu item

image

From v11
image


There are no changes in the migrated code so I'm confused on why this is not working as expected? @max3903 Is there a v12 change in Odoo we need to account for in the migration?

**Review by brian10048 — CHANGES_REQUESTED** [review]
https://github.com/OCA/field-service/pull/251#issuecomment-506351395

https://github.com/OCA/field-service/pull/251#issuecomment-506353954

File: fieldservice_recurring/security/recurring_security.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo noupdate="1">
+
+    <record id="rule_fsm_frequency_multi_company" model="ir.rule">
+        <field name="name">FSM Frequency multi-company</field>
+        <field name="model_id" ref="model_fsm_frequency"/>
+        <field name="global" eval="True"/>
+        <field name="domain_force">['|', ('company_id', '=', False), ('company_id', 'child_of', [user.company_id.id])]</field>
```
Comment (mourad-ehm) [inline fieldservice_recurring/security/recurring_security.xml:8]: HI @kittiu, thanks very much for your contribution. 
I have this bug on fms_frequency. 
Traceback (most recent call last):
  File "/odoo/src/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/odoo/src/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/odoo/src/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/odoo/src/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/odoo/src/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/odoo/src/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/odoo/src/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/odoo/src/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/odoo/src/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/odoo/src/addons/web/controllers/main.py", line 904, in search_read
    return self.do_search_read(model, fields, offset, limit, domain, sort)
  File "/odoo/src/addons/web/controllers/main.py", line 926, in do_search_read
    offset=offset or 0, limit=limit or False, order=sort or False)
  File "/odoo/src/odoo/models.py", line 4572, in search_read
    records = self.search(domain or [], offset=offset, limit=limit, order=order)
  File "/odoo/src/odoo/models.py", line 1561, in search
    res = self._search(args, offset=offset, limit=limit, order=order, count=count)
  File "/odoo/src/odoo/models.py", line 4105, in _search
    self._apply_ir_rules(query, 'read')
  File "/odoo/src/odoo/models.py", line 3957, in _apply_ir_rules
    where_clause, where_params, tables = Rule.domain_get(self._name, mode)
  File "/odoo/src/odoo/addons/base/models/ir_rule.py", line 126, in domain_get
    query = self.env[model_name].sudo()._where_calc(dom, active_test=False)
  File "/odoo/src/odoo/models.py", line 3896, in _where_calc
    e = expression.expression(domain, self)
  File "/odoo/src/odoo/osv/expression.py", line 672, in __init__
    self.parse()
  File "/odoo/src/odoo/osv/expression.py", line 853, in parse
    raise ValueError("Invalid field %r in leaf %r" % (left, str(leaf)))
 ValueError: Invalid field 'company_id' in leaf "<osv.ExtendedLeaf: ('company_id', '=', False) on fsm_frequency (ctx: )>"

File: fieldservice_recurring/models/fsm_frequency.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        string='Repeat Every', help="The number of intervals between events",
+        default=1, required=True, track_visibility='onchange')
+    interval_type = fields.Selection(
+        FREQUENCY_SELECT, string='Interval Type',
+        required=True, track_visibility='onchange')
+    is_exclusive = fields.Boolean(
+        string='Exclusive Rule?', default=False,
+        help="""Checking this box will make this an exclusive rule. Exclusive
+            rules prevent the configured days from being a schedule option""")
+
```
Comment (mourad-ehm) [inline fieldservice_recurring/models/fsm_frequency.py]:
```suggestion
    company_id = fields.Many2one(
        'res.company', 'Company',
        default=lambda self: self.env.user.company_id)

```

File: fieldservice_recurring/security/ir.model.access.csv [hunk 3]
```diff
@@ -0,0 +1,7 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_fsm_frequency_fsm_user,fsm.frequency.user,model_fsm_frequency,fieldservice.group_fsm_user,1,1,1,0
+access_fsm_frequency_fsm_manager,fsm.frequency.manager,model_fsm_frequency,fieldservice.group_fsm_manager,1,1,1,1
+access_fsm_recurring_template_fsm_user,fsm.recurring.template.user,model_fsm_recurring_template,fieldservice.group_fsm_user,1,0,0,0
+access_fsm_recurring_template_fsm_manager,fsm.recurring.template.manager,model_fsm_recurring_template,fieldservice.group_fsm_manager,1,1,1,1
+access_fsm_recurring_fsm_user,fsm.recurring.user,model_fsm_recurring,fieldservice.group_fsm_user,1,1,1,0
+access_fsm_recurring_fsm_manager,fsm.recurring.manager,model_fsm_recurring,fieldservice.group_fsm_manager,1,1,1,1
```
Comment (mourad-ehm) [inline fieldservice_recurring/security/ir.model.access.csv:9]:
```suggestion
access_fsm_recurring_fsm_manager,fsm.recurring.manager,model_fsm_recurring,fieldservice.group_fsm_manager,1,1,1,1
access_fsm_frequency_fsm_set_user,fsm.frequency.set.user,model_fsm_frequency_set,fieldservice.group_fsm_user,1,1,1,0
access_fsm_frequency_fsm_set_manager,fsm.frequency.set.manager,model_fsm_frequency_set,fieldservice.group_fsm_manager,1,1,1,1
```

Comment (mourad-ehm) [reply fieldservice_recurring/security/ir.model.access.csv:9]: That will fix bug of menu Fsm frequency set

File: fieldservice_recurring/models/fsm_recurring.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        ], offset=0, limit=1, order='scheduled_date_start desc')
+        if last_order:
+            next_date = fields.Datetime.from_string(
+                last_order.scheduled_date_start)
+        # set thru_date to use as rrule 'until' parameter
+        days_ahead = self.fsm_frequency_set_id.schedule_days
+        request_thru_date = datetime.now() + relativedelta(days=+days_ahead)
+        request_thru_str = request_thru_date.strftime(
+            DEFAULT_SERVER_DATETIME_FORMAT)
+        if self.end_date and (self.end_date < request_thru_str):
```
Comment (mourad-ehm) [inline fieldservice_recurring/models/fsm_recurring.py]: bug when executing cron.
TypeError: unorderable types: datetime.datetime() < str()

Comment (mourad-ehm) [inline fieldservice_recurring/models/fsm_recurring.py]:
```suggestion
        if self.end_date and (self.end_date < request_thru_date):
```

**Comment by kittiu** [comment]
@brian10048 @mourad-ehm thank you for the review. I will get back to this shortly.

File: fieldservice_recurring/models/fsm_recurring.py (hunk 4 above)
Comment (kittiu) [reply fieldservice_recurring/models/fsm_recurring.py]: Fixed, to become like this
```
        if self.end_date and (self.end_date < request_thru_date):
            thru_date = self.end_date   # Here too, remove the from_string() function
        else:
            thru_date = request_thru_date
```

**Comment by kittiu** [comment]
@brian10048 @mourad-ehm, I have fixed all mentioned. But I will still continue to do test script later, so keep WIP

File: fieldservice_recurring/models/fsm_recurring.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        orders = self.env['fsm.order']
+        for rec in self.env['fsm.recurring'].search([
+            ('state', 'in', ('progress', 'pending'))
+        ]):
+            schedule_dates = rec._get_rruleset()
+            order_dates = []
+            for order in rec.fsm_order_ids:
+                if order.scheduled_date_start:
+                    order_dates.append(
+                        datetime.strptime(order.scheduled_date_start,
```
Comment (brian10048) [inline fieldservice_recurring/models/fsm_recurring.py]: Error here

 ` File "/home/odoo/build/OCA/field-service/fieldservice_recurring/models/fsm_recurring.py", line 184, in _cron_generate_orders
    DEFAULT_SERVER_DATETIME_FORMAT
ValueError: <class 'TypeError'>: "strptime() argument 1 must be str, not datetime.datetime" while evaluating
'model._cron_scheduled_task()'`

Comment (kittiu) [reply fieldservice_recurring/models/fsm_recurring.py]: Fixed.

**Comment by kittiu** [comment]
@brian10048 I read through the code and understand now the concept. If you could give me some good/complex case of rules and the expected when each work order will be create, that would be great.
But if you are busy, it is fine, I still can come out with something.

**Comment by babatoko** [comment]
@kittiu

Here some simple case of recurrent work

- Work order Monday to Friday 6am to 7am and each odd week Wednesday not worked.

- Worked to be done every 3 weeks.

- work only the last day of the month only ending by 31th.

**Comment by kittiu** [comment]
Thank you @babatoko

**Comment by kittiu** [comment]
> @kittiu
> 
> Here some simple case of recurrent work
> 
> * Work order Monday to Friday 6am to 7am and each odd week Wednesday not worked.
> * Worked to be done every 3 weeks.
> * work only the last day of the month only ending by 31th.

Finished all 3 test cases

Note: but I still not quite understand when you say **6am to 7am**, I though the recurring just create a draft order use same time as in start date.

**Comment by kittiu** [comment]
Test added, this module migration is ready for review.
@babatoko @brian10048

By the way, I found that all other field-service modules do not have company_id, and so I have comment this is avoid log warn message,
https://github.com/ecosoft-odoo/field-service/blob/12.0-mig-fieldservice_recurring/fieldservice_recurring/models/fsm_recurring.py#L156

Only in this field_recurring's module has company_id field. Should I remove them all?

**Review by brian10048 — APPROVED** [review]

**Comment by brian10048** [comment]
> Note: but I still not quite understand when you say **6am to 7am**, I though the recurring just create a draft order use same time as in start date.

This module does not currently have functionality to support time windows.  Would be a great addition



> By the way, I found that all other field-service modules do not have company_id, and so I have comment this is avoid log warn message,
> https://github.com/ecosoft-odoo/field-service/blob/12.0-mig-fieldservice_recurring/fieldservice_recurring/models/fsm_recurring.py#L156
> 
> Only in this field_recurring's module has company_id field. Should I remove them all?

In v11 company_id was supported in the base field service modules.  It appears they did not stay in the migration to v12.  @max3903, I'm not sure if this was intentional or not?


I have tested this and it appears to be working as intended

**Comment by max3903** [comment]
@kittiu Can you fix the Travis error please?
https://travis-ci.org/OCA/field-service/jobs/556168537#L1165

**Comment by kittiu** [comment]
> @kittiu Can you fix the Travis error please?
> https://travis-ci.org/OCA/field-service/jobs/556168537#L1165

Fixed

**Comment by max3903** [comment]
Ping @sbidoul

**Pull request merged**